sys-utils/blkreport.8
.TH BLKREPORT 8 "February 2017" "util-linux" "System Administration"
.SH NAME
blkreport \- report zones on a device
.SH SYNOPSIS
.B blkreport
[options]
.RB [ \-z
.IR zone ]
.RB [ \-c
.IR count ]
.I device
.SH DESCRIPTION
.B blkreport
is used to report device zone information.  This is useful for
devices that support the Zoned Block Commands (ZBC) or Zoned-device ATA
Commands (ZAC).
.PP
By default,
.B blkreport
will report on up to 4k zones from the start of the block device.
Options may be used to modify this behavior, changing the starting zone or
the size of the report, as explained below.
.PP
The
.I device
argument is the pathname of the block device.
.SH OPTIONS
The
.I offset
and
.I length
option arguments may be followed by the multiplicative suffixes KiB (=1024),
MiB (=1024*1024), and so on for GiB, TiB, PiB, EiB, ZiB and YiB (the "iB" is
optional, e.g., "K" has the same meaning as "KiB") or the suffixes
KB (=1000), MB (=1000*1000), and so on for GB, TB, PB, EB, ZB and YB.
Additionally, the 0x prefix can be used to specify \fIoffset\fR and
\fIlength\fR in hex.
.TP
.BR \-z , " \-\-zone "\fIoffset\fP
The starting zone of the report specified as a sector offset.
The provided offset in sector units (512 bytes) should match the start of a zone.
The default value is zero.
.TP
.BR \-c , " \-\-count "\fIlength\fP
The maximum number of zones to be returned by the report.
The default is 4096, the maximum 65536.
.TP
.BR \-v , " \-\-verbose"
Display the number of zones returned in the report.
.TP
.BR \-V , " \-\-version"
Display version information and exit.
.TP
.BR \-h , " \-\-help"
Display help text and exit.
.SH REPORT OUTPUT
.TS
tab(:);
left l l.
start:Zone start sector
len:Zone length in number of sectors
wptr:Zone write pointer position
reset:Reset write pointer recommended
non-seq:Non-sequential write resources active
cond:Zone condition
type:Zone type
.TE
.SS Zone conditions
.TS
tab(:);
left l l.
Cl:closed
cv:conventional zone
e0:empty
fu:full
Oe:open explicit
Oi:open implicit
OL:offline
ro:read only
x?:reserved conditions (should not be reported)
.TE
.SH AUTHOR
.MT shaun@tancheff.com
Shaun Tancheff
.ME
.SH SEE ALSO
.BR sg_rep_zones (8)
.SH AVAILABILITY
The blkreport command is part of the util-linux package and is available from
.UR https://\:www.kernel.org\:/pub\:/linux\:/utils\:/util-linux/
Linux Kernel Archive
.UE .


NEWS
util-linux 2.29.1: Jan 20 2017
* see Documentation/releases/v2.29.1-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.29/v2.29.1-ChangeLog

util-linux 2.29: Nov 8 2016
* see Documentation/releases/v2.29-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.29/v2.29-ChangeLog

util-linux 2.29-rc2: Oct 19 2016
* see Documentation/releases/v2.29-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.29/v2.29-rc2-ChangeLog

util-linux 2.29-rc1: Sep 30 2016
* see Documentation/releases/v2.29-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.29/v2.29-rc1-ChangeLog

util-linux 2.28: Apr 12 2016
* see Documentation/releases/v2.28-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.28/v2.28-ChangeLog

util-linux 2.28-rc2: Mar 29 2016
* see Documentation/releases/v2.28-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.28/v2.28-rc2-ChangeLog

util-linux 2.28-rc1: Mar 11 2016
* see Documentation/releases/v2.28-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.28/v2.28-rc1-ChangeLog

util-linux 2.27: Sep 07 2015
* see Documentation/releases/v2.27-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.27/v2.27-ChangeLog

util-linux 2.27-rc2: Aug 24 2015
* see Documentation/releases/v2.27-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.27/v2.27-rc2-ChangeLog

util-linux 2.27-rc1: Jul 31 2015
* see Documentation/releases/v2.27-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.27/v2.27-rc1-ChangeLog

util-linux 2.26: Feb 19 2015
* see Documentation/releases/v2.26-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.26/v2.26-ChangeLog

util-linux 2.26-rc2: Feb 4 2015
* see Documentation/releases/v2.26-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.26/v2.26-rc2-ChangeLog

util-linux 2.26-rc1: Jan 14 2015
* see Documentation/releases/v2.26-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.26/v2.26-rc1-ChangeLog

util-linux 2.25: Jul 22 2014
* see Documentation/releases/v2.25-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.25/v2.25-ChangeLog

util-linux 2.25-rc2: Jul 2 2014
* see Documentation/releases/v2.25-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.25/v2.25-rc2-ChangeLog

util-linux 2.25-rc1: Jun 18 2014
* see Documentation/releases/v2.25-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.25/v2.25-rc1-ChangeLog

util-linux 2.24: Oct 21 2013
* see Documentation/releases/v2.24-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.24/v2.24-ChangeLog

util-linux 2.24-rc2: Oct 11 2013
* see Documentation/releases/v2.24-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.24/v2.24-rc2-ChangeLog

util-linux 2.24-rc1: Sep 27 2013
* see Documentation/releases/v2.24-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.24/v2.24-rc1-ChangeLog

util-linux 2.23: Apr 25 2013
* see Documentation/releases/v2.23-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.23/v2.23-ChangeLog

util-linux 2.23-rc2: Apr 10 2013
* see Documentation/releases/v2.23-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.23/v2.23-rc2-ChangeLog

util-linux 2.23-rc1: Mar 22 2013
* see Documentation/releases/v2.23-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.23/v2.23-rc1-ChangeLog

util-linux 2.22: Sep 04 2012
* see Documentation/releases/v2.22-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.22/v2.22-ChangeLog

util-linux 2.22-rc2: Aug 15 2012
* see Documentation/releases/v2.22-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.22/v2.22-rc2-ChangeLog

util-linux 2.22-rc1: Jul 27 2012
* see Documentation/releases/v2.22-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.22/v2.22-rc1-ChangeLog

util-linux 2.21: Feb 24 2012
* see Documentation/releases/v2.21-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.21/v2.21-ChangeLog

util-linux 2.21-rc2: Feb 06 2012
* see Documentation/releases/v2.21-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.21/v2.21-rc2-ChangeLog

util-linux 2.21-rc1: Jan 18 2012
* see Documentation/releases/v2.21-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.21/v2.21-rc1-ChangeLog

util-linux 2.20: Aug 29 2011
* see Documentation/releases/v2.20-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.20/v2.20-ChangeLog

util-linux 2.20-rc2: Aug 17 2011
* see Documentation/releases/v2.20-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.20/v2.20-rc2-ChangeLog

util-linux 2.20-rc1: Jul 29 2011
* see Documentation/releases/v2.20-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.20/v2.20-rc1-ChangeLog

util-linux 2.19: Feb 10 2011
* see Documentation/releases/v2.19-ReleaseNotes or the complete changelog at
  ftp://ftp.kernel.org/pub/linux/utils/util-linux/v2.19/v2.19-ChangeLog

util-linux 2.19-rc3: Jan 25 2011

sys-utils/blkreport.c
/*
 * blkreport.c -- request a zone report on part (or all) of the block device.
 *
 * Copyright (C) 2015,2016 Seagate Technology PLC
 * Written by Shaun Tancheff <shaun.tancheff@seagate.com>
 *
 * Copyright (C) 2017 Karel Zak <kzak@redhat.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * This program uses BLKREPORT ioctl to query zone information about part of
 * or a whole block device, if the device supports it.
 * You can specify range (start and length) to be queried.
 */

#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <fcntl.h>
#include <limits.h>
#include <getopt.h>
#include <time.h>

#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <linux/fs.h>
#include <linux/blkzoned.h>

#include "nls.h"
#include "strutils.h"
#include "xalloc.h"
#include "c.h"
#include "closestream.h"
#include "blkdev.h"

static const char *type_text[] = {
	"RESERVED",
	"CONVENTIONAL",
	"SEQ_WRITE_REQUIRED",
	"SEQ_WRITE_PREFERRED",
};

static const char *condition_str[] = {
	"cv", /* conventional zone */
	"e0", /* empty */
	"Oi", /* open implicit */
	"Oe", /* open explicit */
	"Cl", /* closed */
	"x5", "x6", "x7", "x8", "x9", "xA", "xB", /* xN: reserved */
	"ro", /* read only */
	"fu", /* full */
	"OL"  /* offline */
};

static void print_zones(const struct blk_zone *info, const uint32_t count)
{
	uint32_t iter;

	printf(_("Zones returned: %u\n"), count);

	for (iter = 0; iter < count; iter++) {
		const struct blk_zone *entry = &info[iter];
		unsigned int type = entry->type;
		uint64_t start = entry->start;
		uint64_t wp = entry->wp;
		uint8_t cond = entry->cond;
		uint64_t len = entry->len;

		if (!len)
			break;

		printf(_("  start: %9lx, len %6lx, wptr %6lx"
			 " reset:%u non-seq:%u, zcond:%2u(%s) [type: %u(%s)]\n"),
			start, len, wp - start,
			entry->reset, entry->non_seq,
			cond, condition_str[cond & ARRAY_SIZE(condition_str)],
			type, type_text[type]);
	}
}

static int do_report(const int fd, const uint64_t lba, uint32_t len, const int verbose)
{
	int rc;
	struct blk_zone_report *zi;

	zi = xmalloc(sizeof(struct blk_zone_report) + (len * sizeof(struct blk_zone)));
	zi->nr_zones = len;
	zi->sector = lba; /* maybe shift 4Kn -> 512e */
	rc = ioctl(fd, BLKREPORTZONE, zi);
	if (rc != -1) {
		if (verbose)
			printf(_("Found %d zones\n"), zi->nr_zones);
		print_zones(zi->zones, zi->nr_zones);
	}
	free(zi);

	return rc;
}

static void __attribute__((__noreturn__)) usage(FILE *out)
{
	fputs(USAGE_HEADER, out);
	fprintf(out, _(" %s [options] <device>\n"), program_invocation_short_name);

	fputs(USAGE_SEPARATOR, out);
	fputs(_("Report zone information about the given device.\n"), out);

	fputs(USAGE_OPTIONS, out);
	fputs(_(" -z, --zone <offset>    start sector of first zone (in 512-byte sectors)\n"), out);
	fputs(_(" -c, --count <number>   maximum number of zones in the report\n"), out);
	fputs(_(" -v, --verbose          display the number of reported zones"), out);
	fputs(USAGE_SEPARATOR, out);
	fputs(USAGE_HELP, out);
	fputs(USAGE_VERSION, out);

	fprintf(out, USAGE_MAN_TAIL("blkreport(8)"));
	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

#define DEF_REPORT_LEN		(1 << 12) /* 4k zones (256k kzalloc) */
#define MAX_REPORT_LEN		(1 << 16) /* 64k zones */

int main(int argc, char **argv)
{
	char *path;
	int c;
	int fd;
	int secsize;
	uint64_t blksectors = 0;
	struct stat sb;
	int verbose = 0;
	uint64_t offset = 0ul;
	uint32_t length = DEF_REPORT_LEN;
	static const struct option longopts[] = {
	    { "help",    no_argument,       NULL, 'h' },
	    { "version", no_argument,       NULL, 'V' },
	    { "zone",    required_argument, NULL, 'z' }, /* starting LBA */
	    { "count",   required_argument, NULL, 'c' }, /* max #of zones (entries) for result */
	    { "verbose", no_argument,       NULL, 'v' },
	    { NULL, 0, NULL, 0 }
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	atexit(close_stdout);

	while ((c = getopt_long(argc, argv, "hc:z:vV", longopts, NULL)) != -1) {
		switch (c) {
		case 'h':
			usage(stdout);
			break;
		case 'c':
			length = strtosize_or_err(optarg,
					_("failed to parse number of zones"));
			break;
		case 'z':
			offset = strtosize_or_err(optarg,
					_("failed to parse zone offset"));
			break;
		case 'v':
			verbose = 1;
			break;
		case 'V':
			printf(UTIL_LINUX_VERSION);
			return EXIT_SUCCESS;
		default:
			usage(stderr);
			break;
		}
	}

	if (optind == argc)
		errx(EXIT_FAILURE, _("no device specified"));

	path = argv[optind++];

	if (optind != argc) {
		warnx(_("unexpected number of arguments"));
		usage(stderr);
	}

	fd = open(path, O_RDONLY);
	if (fd < 0)
		err(EXIT_FAILURE, _("cannot open %s"), path);

	if (fstat(fd, &sb) == -1)
		err(EXIT_FAILURE, _("stat of %s failed"), path);
	if (!S_ISBLK(sb.st_mode))
		errx(EXIT_FAILURE, _("%s: not a block device"), path);

	if (blkdev_get_sectors(fd, (unsigned long long *) &blksectors))
		err(EXIT_FAILURE, _("%s: blkdev_get_sectors ioctl failed"), path);
	if (blkdev_get_sector_size(fd, &secsize))
		err(EXIT_FAILURE, _("%s: BLKSSZGET ioctl failed"), path);

	if (offset > blksectors)
		errx(EXIT_FAILURE, _("%s: offset is greater than device size"), path);
	if (length < 1)
		length = 1;
	if (length > MAX_REPORT_LEN) {
		length = MAX_REPORT_LEN;
		warnx(_("limiting report to %u entries"), length);
	}

	if (do_report(fd, offset, length, verbose))
		 err(EXIT_FAILURE, _("%s: BLKREPORTZONE ioctl failed"), path);

	close(fd);
	return EXIT_SUCCESS;
}
